### PR TITLE
Implement Ardo 4.0 roadmap milestones

### PR DESCRIPTION
**Pull request by swernerx**
## Summary

Implements the Ardo 4.0 roadmap milestone plan from `docs/superpowers/specs/2026-07-09-ardo-4-roadmap-milestone-plan.md`.

## Milestone progress

- [x] Add milestone plan document
- [x] M0: Route Identity Contract
  - [x] Add internal route identity/public URL builder with version and future locale path tests
  - [x] Feed route identity through route manifest, sitemap, redirects, llms output, and link checking
  - [x] Add version/locale/public path metadata to current search index
  - [x] Prepare generated sidebar data for route identity/public paths
- [x] M1: Versioning Stabilization
  - [x] Preserve equivalent route paths in the version switcher, including deployment-base paths
  - [x] Keep root redirects, versions.json, sitemap, search, sidebars, redirects, and link checks on route identity/public paths
- [x] M2: i18n URL Preparation
  - [x] Add the reserved `i18n` config contract with validation and label defaults
  - [x] Feed default locale into route identity, manifest, search, sidebars, sitemap, and llms paths when i18n is active
  - [x] Redirect `/` to the current version plus default locale and redirect the version root to the default locale root
  - [x] Expose `versioning` and `i18n` in the virtual runtime config
  - [x] Document the URL-prep-only i18n contract
- [x] M3: Metadata Pipeline
  - [x] Add normalized `PageMetadata` and `PageFrontmatterMetadata` models
  - [x] Attach normalized metadata to route manifest entries
  - [x] Move search index generation onto route manifest/page metadata
  - [x] Switch sitemap, redirects, and llms output to metadata fields
  - [x] Share frontmatter metadata parsing with sidebar generation
- [x] M4: Search 2.0
  - [x] Generate section-level records with page title, heading hierarchy, anchors, excerpts, version/locale metadata, route group, and public URL
  - [x] Keep the virtual search module compatible with the current MiniSearch UI
  - [x] Emit static self-hostable `search/manifest.json` and `search/chunk-0.json` build assets
  - [x] Add tests for anchor records, duplicate heading slugs, version/locale metadata, static assets, and UI navigation
- [x] M5: Content Source Mapping
  - [x] Add narrow `content` source mapping on the Vite plugin
  - [x] Materialize external Markdown/MDX into generated route directories with `.ardo-generated` safety markers
  - [x] Synthesize default title/order metadata and support a small frontmatter override hook
  - [x] Feed generated files through the existing route manifest and metadata pipeline
  - [x] Document the interim scope and single-file `index.md` materialization behavior
- [x] M6: Extension API Preparation
  - [x] Add internal lifecycle phase registry and runner
  - [x] Thread content source materialization, route generation, metadata scan, and build output emission through named phases
  - [x] Document which phase names are internal and which are future public candidates
  - [x] Add lifecycle tests without exposing a public extension API
- [x] Runtime compatibility follow-up
  - [x] Keep deprecated runtime `generateSidebar()` output stable while Vite-generated sidebars carry route identity metadata
- [x] E2E Search 2.0 follow-up
  - [x] Update the docs smoke test to expect section-level search navigation to `/v3/guide/configuration#markdown`

## Validation

- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/route-identity.test.ts packages/ardo/src/vite/route-manifest.test.ts packages/ardo/src/vite/build-outputs.test.ts packages/ardo/src/vite/llms-text.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/search-index.test.ts packages/ardo/src/vite/route-identity.test.ts packages/ardo/src/vite/route-manifest.test.ts packages/ardo/src/vite/build-outputs.test.ts packages/ardo/src/vite/llms-text.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/sidebar-index.test.ts packages/ardo/src/vite/search-index.test.ts packages/ardo/src/vite/route-identity.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/ui/shell.test.tsx
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/config/index.test.ts packages/ardo/src/vite/route-identity.test.ts packages/ardo/src/vite/build-outputs.test.ts packages/ardo/src/vite/plugin.test.ts packages/ardo/src/vite/sidebar-index.test.ts packages/ardo/src/vite/search-index.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/route-manifest.test.ts packages/ardo/src/vite/search-index.test.ts packages/ardo/src/vite/build-outputs.test.ts packages/ardo/src/vite/llms-text.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/route-manifest.test.ts packages/ardo/src/vite/sidebar-index.test.ts packages/ardo/src/vite/search-index.test.ts packages/ardo/src/vite/build-outputs.test.ts packages/ardo/src/vite/llms-text.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/search-index.test.ts packages/ardo/src/vite/build-outputs.test.ts packages/ardo/src/ui/components/Search.interaction.test.tsx
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/content-sources.test.ts packages/ardo/src/vite/route-manifest.test.ts packages/ardo/src/vite/plugin.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/vite/lifecycle.test.ts packages/ardo/src/vite/content-sources.test.ts packages/ardo/src/vite/plugin.test.ts packages/ardo/src/vite/build-outputs.test.ts
- [x] CI=true pnpm exec vitest --run --config vitest.config.ts packages/ardo/src/runtime/sidebar.test.ts packages/ardo/src/vite/sidebar-index.test.ts
- [x] CI=true pnpm --filter ardo typecheck
- [x] Targeted ESLint for changed M0/M1/M2/M3/M4/M5/M6 package files
- [x] CI=true pnpm exec prettier --check docs/app/routes/guide/configuration.mdx
- [x] CI=true pnpm exec prettier --check tests/e2e/docs-smoke.spec.ts
- [x] CI=true pnpm build
- [x] CI=true pnpm test
- [x] CI=true pnpm typecheck
- [x] CI=true pnpm lint (warnings only, existing warning budget)
- [x] CI=true pnpm docs:build
- [x] CI=true pnpm test:e2e
- [x] Pre-push hook: pnpm -r typecheck
- [x] Pre-push hook: eslint packages/*/src (warnings only, existing warning budget)
- [x] GitHub Actions PR matrix: Test (Node.js 22), Test (Node.js 24), Test (Node.js 26)
- [x] Full repo validation before marking ready for review

### CODE REVIEW

File: packages/ardo/src/vite/search-index.ts [hunk 1]
```diff
@@ -1,104 +1,217 @@
...
+  const slugger = createHeadingSlugger()
+  const sections: MarkdownSection[] = [{ content: "", headingHierarchy: [] }]
+  const headingStack: Array<{ level: number; title: string }> = []
+
+  for (const line of content.split("\n")) {
+    const heading = parseMarkdownHeading(line)
+    if (heading == null) {
+      appendSectionContentLine(sections, line)
+      continue
+    }
```
Comment (swernerx) [inline packages/ardo/src/vite/search-index.ts:132]: `splitMarkdownSections` treats every `# …` line as a heading, **including lines inside fenced code blocks** — so a sample like `# install deps` inside a code fence becomes its own search record with an anchor that rehype-slug never emits on the rendered page (clicking the result scrolls nowhere). `sanitizeSearchContent` already strips fences a few lines below; mirroring that here keeps section splitting and the real page anchors in sync:

````suggestion
  let isInsideFence = false
  for (const line of content.split("\n")) {
    if (line.trimStart().startsWith("```")) {
      isInsideFence = !isInsideFence
      appendSectionContentLine(sections, line)
      continue
    }

    const heading = isInsideFence ? null : parseMarkdownHeading(line)
    if (heading == null) {
      appendSectionContentLine(sections, line)
      continue
    }
````

Heads-up: `extractAnchors` in `route-manifest.ts` has the same blind spot (pre-existing), so it's worth fixing both for consistency between link-checking and search.

File: packages/ardo/src/vite/build-versioning.ts [hunk 2]
```diff
@@ -41,5 +43,42 @@ export function getRootVersionRedirect(
...
+
+export function getVersioningRedirects(
+  config: ResolvedConfig
+): Array<{ from: string; to: string }> {
+  return [getRootVersionRedirect(config), getVersionRootLocaleRedirect(config)].filter(
+    (redirect): redirect is { from: string; to: string } => redirect != null
+  )
+}
+
+export function getVersionRootLocaleRedirect(
```
Comment (swernerx) [inline packages/ardo/src/vite/build-versioning.ts:64]: This (and `getRootVersionRedirect` above) emit redirects into `/…/<locale>/`, and `generateSitemap` advertises the matching locale-prefixed `publicPath`s. But the served base from `resolveVersionedBase()` never includes the locale — the build outputs pages at `/docs/v3/guide`, not `/docs/v3/en/guide`. So the moment someone sets `i18n`, `/` → `/docs/v3/en/` and every sitemap `<loc>` resolve to 404s.

Given 4.0 i18n is URL-prep only, I'd suggest one of:
- gate `getVersionRootLocaleRedirect` + the locale segment in `publicPath` behind an actual "locale content is served" flag, so nothing advertises URLs that don't exist yet; **or**
- keep the reservation but make `configuration.mdx` and the `i18n` TSDoc explicit that enabling it currently produces non-resolving canonical URLs (not production-ready).

Either way, a documented option shouldn't silently ship broken redirects + sitemap.

File: packages/ardo/src/vite/content-sources-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    configureServer(server) {
+      for (const source of sources) {
+        server.watcher.add(path.resolve(root, source.from))
+      }
+
+      const handleChange = (changedPath: string) => {
+        if (shouldHandleSourceChange(changedPath, root, sources)) {
+          void materializeAndWriteRoutes()
+        }
+      }
```
Comment (swernerx) [inline packages/ardo/src/vite/content-sources-plugin.ts:78]: `materializeAndWriteRoutes()` deletes+recreates the whole target dir and rewrites `app/routes.ts` on every `.md(x)` add/change/unlink, with no debounce and no guard against overlapping runs. Two small changes make this safer for the dev server:

- **Serialize concurrent runs** — a burst of watcher events can currently start two `rm -rf` + rewrite cycles at once. A module-level in-flight promise you chain/skip on avoids the race:
  ```ts
  let inFlight: Promise<void> | null = null
  const handleChange = (changedPath: string) => {
    if (!shouldHandleSourceChange(changedPath, root, sources)) return
    inFlight = (inFlight ?? Promise.resolve()).then(() => materializeAndWriteRoutes())
  }
  ```
- **Only rewrite `app/routes.ts` when the generated file set actually changed**, so saving a source file doesn't trigger a full reload on every keystroke-save.

Not blocking, but it'll churn noticeably on large ADR/changelog folders.

File: packages/ardo/src/vite/build-outputs.ts [hunk 4]
```diff
@@ -68,14 +69,11 @@ export function generateRobots(config: ResolvedConfig) {
...
-  if (rootVersionRedirect != null) {
-    redirects.push(rootVersionRedirect)
-  }
+  redirects.push(...getVersioningRedirects(config))
 
   for (const entry of entries) {
-    for (const from of entry.frontmatter.redirectFrom ?? []) {
-      redirects.push({ from, to: entry.path })
+    for (const from of entry.metadata.redirectFrom ?? []) {
+      redirects.push({ from, to: entry.publicPath })
```
Comment (swernerx) [inline packages/ardo/src/vite/build-outputs.ts:78]: Switching `to` to `entry.publicPath` (base-prefixed) while leaving `from` as the raw authored value creates an asymmetry: with `base: "/docs/"` this emits `/old-guide → /docs/guide`. For the GitHub-Pages HTML redirect files that's arguably right (the file is placed relative to deploy root, and the meta-refresh needs the full path). But the same pair also lands in `_redirects` / `vercel.json`, where the platform matches `from` at the domain root — there `/old-guide` won't match a site mounted under `/docs/`. Worth either prefixing `from` for the platform-redirect outputs, or adding a comment explaining why the two sides are intentionally different.

File: packages/ardo/src/vite/search-index.ts [hunk 5]
```diff
@@ -1,104 +1,217 @@
...
+    const pathWithAnchor = appendAnchor(entry.routePath, section.anchor)
+    const publicPathWithAnchor = appendAnchor(entry.publicPath, section.anchor)
+
+    return [
+      {
+        id: `${entry.metadata.sourcePath}#${section.anchor ?? `page-${index}`}`,
+        title,
+        pageTitle,
+        content: joinSearchTextParts([title, pageTitle, ...headingHierarchy, excerpt]),
+        excerpt,
```
Comment (swernerx) [inline packages/ardo/src/vite/search-index.ts:96]: Minor payload note: `content` is already `joinSearchTextParts([title, pageTitle, ...headingHierarchy, excerpt])`, and `excerpt` (up to 2000 chars) is then shipped **again** as its own field in the default-exported `searchDocs`. Since the current UI (`Search.tsx`, `storeFields: ["title","path","section"]`) doesn't render `excerpt` / `pageTitle` / `headingHierarchy` / `routeGroup` yet, the client downloads a meaningfully larger index for no current benefit. Fine for the prep milestone — but the plan calls out "index size is measurable", so a `log()` of the emitted index/chunk size would help catch regressions early.

**Review by swernerx — COMMENTED** [review]
Thanks for the huge, well-structured PR — the M0 route-identity contract and centralizing all URL building in one `buildPublicPath()` are exactly the right foundation, and the test coverage (route-identity, content-sources, lifecycle, section search, i18n redirects, version switching) is excellent.

I left inline comments with concrete suggestions. Two are worth addressing before/soon after merge:

1. **i18n, when enabled, advertises canonical URLs the static build never serves** (highest impact). `publicPath` gets the locale segment, but the deployed base (`resolveVersionedBase`) never does — so sitemap `<loc>`s and the `/` + version-root redirects point at `/v3/en/...` while the pages actually live at `/v3/...`. Enabling the documented `i18n` option therefore 404s the homepage redirect and the entire sitemap. Since 4.0 i18n is "URL-prep only", I'd either hold back the locale redirect/sitemap emission until locale content is really served, or make the docs + TSDoc loudly state it isn't production-ready.

2. **Section search parses `#` headings inside fenced code blocks** → phantom results that link to anchors the rendered page never emits. One-line fix inline.

The rest are smaller quality/robustness notes. None of them block the core versioning/metadata/search work, which looks solid. Details below. 👇

File: packages/ardo/src/vite/build-versioning.ts [hunk 6]
```diff
@@ -41,5 +43,42 @@ export function getRootVersionRedirect(
...
+
+  return {
+    from: currentVersion.path,
+    to: buildPublicPath({
+      basePath: currentVersion.path,
+      localeId: config.i18n.defaultLocale,
+      routePath: "/",
+    }),
+  }
 }
```
[the remark this answers is not in the document]
Comment (swernerx) [reply packages/ardo/src/vite/build-versioning.ts:88]: Fixed in 6a53538. getVersionRootLocaleRedirect now also respects rootRedirect: false, with a build-output regression test covering the i18n case.

File: packages/ardo/src/vite/content-sources-plugin.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      root = userConfig.root ?? root
+      routesDir = resolveRoutesDir(root, options.routesDirOption)
+      await materializeAndWriteRoutes()
+    },
+    configResolved(config) {
+      root = config.root
+      routesDir = resolveRoutesDir(root, options.routesDirOption)
+    },
+    async buildStart() {
+      await materializeAndWriteRoutes()
```
[the remark this answers is not in the document]
Comment (swernerx) [reply packages/ardo/src/vite/content-sources-plugin.ts:63]: Fixed in 6a53538. The config hook materialization now marks the run, and buildStart skips the duplicate work for that build. Added a plugin regression test for this.

File: packages/ardo/src/vite/search-index.ts (hunk 1 above)
Comment (swernerx) [reply packages/ardo/src/vite/search-index.ts:132]: Fixed in 6a53538. Route manifest anchor extraction and search section splitting now share Markdown fence detection for triple-backtick and tilde fences, and tests cover headings inside fenced code blocks.

File: packages/ardo/src/vite/content-sources-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      const handleChange = (changedPath: string) => {
+        if (shouldHandleSourceChange(changedPath, root, sources)) {
+          void materializeAndWriteRoutes()
+        }
+      }
+
+      server.watcher.on("add", handleChange)
+      server.watcher.on("change", handleChange)
+      server.watcher.on("unlink", handleChange)
+    },
```
[the remark this answers is not in the document]
Comment (swernerx) [reply packages/ardo/src/vite/content-sources-plugin.ts:83]: Fixed in 6a53538 by chaining watcher materialization through a queued promise, so bursty HMR events run sequentially instead of overlapping delete/write cycles. Added regression coverage in content-sources-plugin.test.ts.

File: packages/ardo/src/vite/build-versioning.ts (hunk 2 above)
Comment (swernerx) [reply packages/ardo/src/vite/build-versioning.ts:64]: Fixed in 6a53538 by making the Vite served/build base include the default locale when i18n is active, while the resolved Ardo deployment base remains version-root so publicPath appends the locale exactly once. Added a plugin test for /v3/en/.

File: packages/ardo/src/vite/content-sources-plugin.ts (hunk 3 above)
Comment (swernerx) [reply packages/ardo/src/vite/content-sources-plugin.ts:78]: Fixed the race part in 6a53538 by serializing watcher materialization runs and adding coverage for bursty HMR events. app/routes.ts already goes through writeRoutesFile, which skips writing unchanged route content, so I kept that behavior as-is here.

File: packages/ardo/src/vite/build-outputs.ts (hunk 4 above)
Comment (swernerx) [reply packages/ardo/src/vite/build-outputs.ts:78]: Addressed in 6a53538 by documenting the existing contract in collectRedirects: from remains the authored static source path, while to uses the canonical public path. I kept behavior unchanged in this PR rather than changing provider semantics alongside the roadmap work.

File: packages/ardo/src/vite/search-index.ts (hunk 5 above)
Comment (swernerx) [reply packages/ardo/src/vite/search-index.ts:96]: Addressed in 6a53538 by adding byteSize to each search manifest chunk. That makes emitted search index size measurable from the static search/manifest.json without adding noisy build logs.
